### PR TITLE
Raiplayradio

### PR DESCRIPTION
**Pull request by frafra**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [X] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Description of your pull request and other information

Support for www.raiplayradio.it added.

Notes:

- Using `--embed-thumbnail` produces ERROR: Conversion failed! on some links, like the one in the tests: I have no idea why that is happening, but the downloaded temporary image has a .jpg extension but it is encoded in webp (the URL is correct). The resolution and ratio are also different.

Here is an example:

`./yt-dlp --add-metadata -o '%(artist)s/%(album)s/%(title)s.%(ext)s' https://www.raiplayradio.it/playlist/2017/12/Alice-nel-paese-delle-meraviglie-72371d3c-d998-49f3-8860-d168cfdf4966.html`

Other pages to test can be found at: https://www.raiplayradio.it/programmi/adaltavoce/archivio/audiolibri/tutte/1

This is a port of my previous PR on youtube-dl: https://github.com/ytdl-org/youtube-dl/pull/21837

### CODE REVIEW

File: supportedsites.md [hunk 1]
```diff
@@ -816,6 +816,8 @@
  - **RaiPlay**
  - **RaiPlayLive**
  - **RaiPlayPlaylist**
+ - **RaiPlayRadio**
+ - **RaiPlayRadioPlaylist**
```
Comment (pukkandan) [inline supportedsites.md:820]: this will be auto-generated at time of release
```suggestion
```

File: yt_dlp/extractor/rai.py [hunk 2]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+
+    def handle_starttag(self, tag, attrs):
+        if tag == 'li' and self._level == 0:
+            self.items.append(dict(attrs))
+        self._level += 1
+
+    def handle_endtag(self, tag):
+        self._level -= 1
+
+
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:607]: You may want to move this into utils

File: yt_dlp/extractor/rai.py [hunk 3]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+class RaiPlayRadioBaseIE(InfoExtractor):
+    _BASE = 'https://www.raiplayradio.it'
+
+    def parse_list(self, webpage):
+        parser = HTMLListAttrsParser()
+        parser.feed(webpage)
+        parser.close()
+        return parser.items
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:615]: same

File: yt_dlp/extractor/rai.py [hunk 4]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+                'title': title,
+                'ext': 'mp3',
+                'language': 'it',
+            }
+            if 'data-image' in attrs:
+                entry['thumbnail'] = urljoin(url, attrs['data-image'])
+            yield entry
+
+    def get_playlist(self, *args, **kwargs):
+        return list(self.get_playlist_iter(*args, **kwargs))
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:634]: unnecessary. only the iter version is needed

File: yt_dlp/extractor/rai.py [hunk 5]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+    _VALID_URL = r'%s/audio/.+?-(?P<id>%s)\.html' % (
+        RaiPlayRadioBaseIE._BASE, RaiBaseIE._UUID_RE)
+    _TEST = {
+        'url': 'https://www.raiplayradio.it/audio/2019/07/RADIO3---LEZIONI-DI-MUSICA-36b099ff-4123-4443-9bf9-38e43ef5e025.html',
+        'info_dict': {
+            'id': '36b099ff-4123-4443-9bf9-38e43ef5e025',
+            'ext': 'mp3',
+            'title': 'Dal "Chiaro di luna" al  "Clair de lune", '
+                     'prima parte con Giovanni Bietti',
+            'thumbnail': r're:^https?://.*\.jpg$',
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:647]: use md5. Also, generally, don't split strings in multiple lines

File: yt_dlp/extractor/rai.py [hunk 6]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+            'language': 'it',
+        }
+    }
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+        list_url = url.replace('.html', '-list.html')
+        for entry in self.get_playlist_iter(list_url, audio_id):
+            if entry['id'] == audio_id:
+                return entry
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:657]:
```suggestion
        return next(entry for entry in self.get_playlist_iter(list_url, audio_id) if entry['id'] == audio_id)
```

File: yt_dlp/extractor/rai.py [hunk 7]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+            'nomeProgramma', playlist_webpage)
+        playlist_description = get_element_by_class(
+            'textDescriptionProgramma', playlist_webpage)
+
+        player_href = self._html_search_regex(
+            r'data-player-href="(.+?)"', playlist_webpage, 'href')
+        list_url = urljoin(url, player_href)
+
+        entries = self.get_playlist(list_url, playlist_id)
+        for index, entry in enumerate(entries, start=1):
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:688]:
```suggestion
        for index, entry in enumerate(self.get_playlist_iter(list_url, playlist_id), start=1):
```

File: yt_dlp/extractor/rai.py [hunk 8]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+            entry.update({
+                'track': entry['title'],
+                'track_number': index,
+                'artist': playlist_creator,
+            })
+            if playlist_title:
+                entry['album'] = playlist_title
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:695]:
```suggestion
                'album': playlist_title
            })
```
Setting `None` in infodict is considered same as not setting anything

File: yt_dlp/extractor/rai.py [hunk 9]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+
+        return self.playlist_result(
+            entries, playlist_id, playlist_title, playlist_description)
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:698]: You can add more fields to this if needed. Eg:
```py
        return self.playlist_result(
            entries, playlist_id, playlist_title, playlist_description, uploader=playlist_creator)
```

I dont know if it's really applicable here, just letting you know

**Review by pukkandan — CHANGES_REQUESTED** [review]
webp downloading as jpg is a common issue. Generally the embedder should handle it. I will have to look into why it doesn't in this case

btw, the 80 char softlimit isnt really somthing to worry about as long as the code isn't obnoxiously long and is easily readable. (which reminds me that I should update Contributing.md)

**Comment by Ashish0804** [comment]
@frafra any updates?

File: yt_dlp/extractor/rai.py (hunk 5 above)
Comment (pukkandan) [reply yt_dlp/extractor/rai.py:647]:
```suggestion
            'title': 'Dal "Chiaro di luna" al  "Clair de lune", prima parte con Giovanni Bietti',
            'thumbnail': r're:^https?://.*\.jpg$',
```

File: yt_dlp/extractor/rai.py [hunk 10]
```diff
@@ -585,3 +588,111 @@ def _real_extract(self, url):
...
+                'ext': 'mp3',
+                'language': 'it',
+            }
+            if 'data-image' in attrs:
+                entry['thumbnail'] = urljoin(url, attrs['data-image'])
+            yield entry
+
+    def get_playlist(self, *args, **kwargs):
+        return list(self.get_playlist_iter(*args, **kwargs))
+
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:635]:
```suggestion
```

File: yt_dlp/extractor/rai.py [hunk 11]
```diff
@@ -585,3 +588,103 @@ def _real_extract(self, url):
...
+        playlist_creator = self._html_search_meta(
+            'nomeProgramma', playlist_webpage)
+        playlist_description = get_element_by_class(
+            'textDescriptionProgramma', playlist_webpage)
+
+        player_href = self._html_search_regex(
+            r'data-player-href="(.+?)"', playlist_webpage, 'href')
+        list_url = urljoin(url, player_href)
+
+        for index, entry in enumerate(self.get_playlist_iter(list_url, playlist_id), start=1):
```
Comment (pukkandan) [inline yt_dlp/extractor/rai.py:681]:
```suggestion

        entries = list(self.get_playlist_iter(list_url, playlist_id))
        for index, entry in enumerate(entries, start=1):
```

**Comment by frafra** [comment]
> @frafra any updates?

I will try to fix the remaining requested changes during the weekend.

Thanks!

File: yt_dlp/extractor/rai.py (hunk 9 above)
Comment (frafra) [reply yt_dlp/extractor/rai.py]: That is a good idea. I added `creator`.

**Comment by frafra** [comment]
Rebased and updated. Ready! :)

**Pull request merged**